### PR TITLE
CI: Added coverage analyzer for esp_modem host tests

### PR DESCRIPTION
**Pull request by gabsuren**

### CODE REVIEW

File: .github/workflows/gcov_analyzer.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+          cd $GITHUB_WORKSPACE/${{ env.COMP_DIR }}/test/host_test
+          cat sdkconfig.ci.coverage >> sdkconfig.defaults
+          idf.py build
+          ./build/host_modem_test.elf -r junit -o junit.xml
+      - name: Run gcovr
+        shell: bash
+        run: |
+          python -m pip install gcovr
+          cd $GITHUB_WORKSPACE/${{ env.COMP_DIR }}
+          gcov-8 test/host_test/build/esp-idf/esp_modem/CMakeFiles/__idf_esp_modem.dir/src/*
```
Comment (david-cermak) [inline .github/workflows/gcov_analyzer.yml:45]: Question: Could we somehow calculate this path? It seems to me that it's an internally created folder structure by CMake

File: .github/workflows/gcov_analyzer.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+          apt-get update && apt-get install -y gcc-8 g++-8 python3-pip
+          apt-get install -y rsync
+          update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-8 800 --slave /usr/bin/g++ g++ /usr/bin/g++-8
+          export DEBIAN_FRONTEND=noninteractive
+          export LWIP_PATH=`pwd`/${{ env.lwip }}
+          export LWIP_CONTRIB_PATH=`pwd`/${{ env.lwip_contrib }}
+          wget --no-verbose ${lwip_uri}/${lwip}.zip
+          unzip -oq ${lwip}.zip
+          wget --no-verbose ${lwip_uri}/${lwip_contrib}.zip
+          unzip -oq ${lwip_contrib}.zip
```
Comment (david-cermak) [inline .github/workflows/gcov_analyzer.yml:34]: Is this a copy-paste from the host tests? Could we make it a subroutine or something reusable, e.g. a script?

File: .github/workflows/gcov_analyzer.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          . ${IDF_PATH}/export.sh
+          cd $GITHUB_WORKSPACE/${{ env.COMP_DIR }}/test/host_test
+          cat sdkconfig.ci.coverage >> sdkconfig.defaults
+          idf.py build
+          ./build/host_modem_test.elf -r junit -o junit.xml
```
Comment (david-cermak) [inline .github/workflows/gcov_analyzer.yml:39]: Nitpick: no need to export junit report, correct?
```suggestion
          ./build/host_modem_test.elf
```

File: components/esp_modem/CMakeLists.txt [hunk 4]
```diff
@@ -47,3 +47,23 @@ if(${target} STREQUAL "linux")
...
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --coverage -fprofile-arcs -ftest-coverage")
+ 
+    SET(CMAKE_EXE_LINKER_FLAGS_COVERAGE
+        ""
+        CACHE STRING "Flags used for linking binaries during coverage builds."
+        FORCE )
+    SET(CMAKE_SHARED_LINKER_FLAGS_COVERAGE
+        ""
+        CACHE STRING "Flags used by the shared libraries linker during coverage builds."
+        FORCE )
```
Comment (david-cermak) [inline components/esp_modem/CMakeLists.txt:62]: Question: What is the purpose of these flags? Do we have to store these messages in CMakeCache?

File: components/esp_modem/Kconfig [hunk 5]
```diff
@@ -25,4 +25,10 @@ menu "esp-modem"
             The typical reason for failing SABM request without a delay is that
             some devices (SIM800) send MSC requests just after opening a new DLCI.
 
+    config GCOV_ENABLED
+        bool "Coverage analyzer"
+        default n
+        help
+	        Enables coverage analyzing for host tests.
+
```
Comment (david-cermak) [inline components/esp_modem/Kconfig:33]: Is it possible to move these options to host test configuration?

**Comment by david-cermak** [comment]
@gabsuren Nice work! Does this PR provide the coverage summary only or is a detailed report available somewhere?

File: .github/workflows/gcov_analyzer.yml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        with:
+          filename: esp-protocols/**/esp_modem_coverage.xml
+          badge: true
+          fail_below_min: false
+          format: markdown
+          hide_branch_rate: false
+          hide_complexity: false
+          indicators: true
+          output: both
+          thresholds: '60 80'
```
Comment (david-cermak) [inline .github/workflows/gcov_analyzer.yml:59]: Does it mean that the coverage job fails if it's below `60%` ? Tried to rerun the job and it failed.

Comment (gabsuren) [reply .github/workflows/gcov_analyzer.yml:59]: @david-cermak it's strange that it failed as I marked flag `fail_below_min: false`. I re-runed the job and it passed
https://github.com/espressif/esp-protocols/pull/196/files#:~:text=fail_below_min%3A%20false

**Review by david-cermak — APPROVED** [review]

**Comment by david-cermak** [comment]
@gabsuren  CI fails in https://github.com/espressif/esp-protocols/actions/runs/3874082428/workflow

**Pull request merged**